src/RestApi/StoreApi/Utilities/ProductQuery.php

```diff
@@ -0,0 +1,351 @@
+			'sku',
+			'min_price',
+			'max_price',
+			'stock_status',
+		);
+
+		foreach ( $custom_keys as $key ) {
+			if ( ! empty( $request[ $key ] ) ) {
+				$args[ $key ] = $request[ $key ];
+			}
+		}
+
+		// Taxonomy query to filter products by type, category,
+		// tag, shipping class, and attribute.
+		$tax_query = array();
+
+		// Map between taxonomy name and arg's key.
+		$taxonomies = array(
+			'product_cat' => 'category',
+			'product_tag' => 'tag',
+		);
+
+		// Set tax_query for each passed arg.
+		foreach ( $taxonomies as $taxonomy => $key ) {
+			if ( ! empty( $request[ $key ] ) ) {
+				$tax_query[] = array(
+					'taxonomy' => $taxonomy,
+					'field'    => 'term_id',
+					'terms'    => $request[ $key ],
+				);
+			}
+		}
+
+		// Filter product type by slug.
+		if ( ! empty( $request['type'] ) ) {
+			$tax_query[] = array(
+				'taxonomy' => 'product_type',
+				'field'    => 'slug',
+				'terms'    => $request['type'],
+			);
+		}
+
+		// Filter by attributes.
+		if ( ! empty( $request['attributes'] ) ) {
+			foreach ( $request['attributes'] as $attribute ) {
+				if ( empty( $attribute['term_id'] ) && empty( $attribute['slug'] ) ) {
+					continue;
+				}
+				if ( in_array( $attribute['attribute'], wc_get_attribute_taxonomy_names(), true ) ) {
+					$tax_query[] = array(
+						'taxonomy' => $attribute['attribute'],
+						'field'    => ! empty( $attribute['term_id'] ) ? 'term_id' : 'slug',
+						'terms'    => ! empty( $attribute['term_id'] ) ? $attribute['term_id'] : $attribute['slug'],
+						'operator' => isset( $attribute['operator'] ) ? $attribute['operator'] : 'IN',
+					);
+				}
+			}
+		}
+
+		// Build tax_query if taxonomies are set.
+		if ( ! empty( $tax_query ) ) {
+			if ( ! empty( $args['tax_query'] ) ) {
+				$args['tax_query'] = array_merge( $tax_query, $args['tax_query'] ); // phpcs:ignore
+			} else {
+				$args['tax_query'] = $tax_query; // phpcs:ignore
+			}
+		}
+
+		// Filter featured.
+		if ( is_bool( $request['featured'] ) ) {
+			$args['tax_query'][] = array(
+				'taxonomy' => 'product_visibility',
+				'field'    => 'name',
+				'terms'    => 'featured',
+				'operator' => true === $request['featured'] ? 'IN' : 'NOT IN',
+			);
+		}
+
+		// Filter by on sale products.
+		if ( is_bool( $request['on_sale'] ) ) {
+			$on_sale_key = $request['on_sale'] ? 'post__in' : 'post__not_in';
+			$on_sale_ids = wc_get_product_ids_on_sale();
+
+			// Use 0 when there's no on sale products to avoid return all products.
+			$on_sale_ids = empty( $on_sale_ids ) ? array( 0 ) : $on_sale_ids;
+
+			$args[ $on_sale_key ] += $on_sale_ids;
+		}
+
+		$operator_mapping = array(
+			'in'     => 'IN',
+			'not_in' => 'NOT IN',
+			'and'    => 'AND',
+		);
+
+		$category_operator  = $operator_mapping[ $request->get_param( 'category_operator' ) ];
+		$tag_operator       = $operator_mapping[ $request->get_param( 'tag_operator' ) ];
+		$attribute_operator = $operator_mapping[ $request->get_param( 'attribute_operator' ) ];
```

Comment:
will these cause warnings if the param is not in the request?  Should check if exists first?